packages/devtools_app/lib/src/screens/memory/shared/heap/model.dart

```diff
@@ -76,41 +77,69 @@ class AdaptedHeapData {
         _JsonFields.created: created.toIso8601String(),
       };
 
-  HeapPath? _retainingPath(IdentityHashCode code) {
+  int? objectIndexByIdentityHashCode(IdentityHashCode code) =>
+      _objectsByCode[code];
+
+  HeapPath? retainingPath(int objectIndex) {
     assert(isSpanningTreeBuilt);
-    var i = _objectsByCode[code]!;
-    if (objects[i].retainer == null) return null;
 
-    final result = <int>[];
+    if (objects[objectIndex].retainer == null) return null;
+
+    final result = <AdaptedHeapObject>[];
 
-    while (i >= 0) {
-      result.add(i);
-      i = objects[i].retainer!;
+    while (objectIndex >= 0) {
+      final object = objects[objectIndex];
+      result.add(object);
+      objectIndex = object.retainer!;
     }
 
-    return result.reversed.toList(growable: false);
+    return HeapPath(result.reversed.toList(growable: false));
   }
+}
+
+/// Result of invocation of [identityHashCode].
+typedef IdentityHashCode = int;
+
+/// Sequence of ids of objects in the heap.
+///
+/// TODO(polina-c): maybe we do not need to store path by objects.
+/// It can be that only classes are interesting, and we can save some
+/// performance on this object. It will become clear when the leak tracking
+/// feature stabilizes.
+class HeapPath {
+  HeapPath(this.objects);
+
+  final List<AdaptedHeapObject> objects;
 
   /// Retaining path for the object in string format.
-  String? shortPath(IdentityHashCode code) {
-    final path = _retainingPath(code);
-    if (path == null) return null;
-    return '/${path.map((i) => objects[i].shortName).join('/')}/';
-  }
+  String? shortPath() => '/${objects.map((o) => o.shortName).join('/')}/';
 
   /// Retaining path for the object as an array of the retaining objects.
-  List<String>? detailedPath(IdentityHashCode code) {
-    final path = _retainingPath(code);
-    if (path == null) return null;
-    return path.map((i) => objects[i].name).toList();
-  }
+  List<String>? detailedPath() =>
```

Comment:
do we have the source locations for each part of this path? if so, would be great to have each part of the path be a link that will open that line of code in the IDE or in the debugger code view. Not for this PR, but an idea for the future.

Reply:
It would be nice, but heap snapshot does not seem to contain location.